src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_x509store.c

```diff
@@ -0,0 +1,375 @@
+
+    char buffer[9] = {0}; // 8-character hex + null terminator
+    size_t written = (size_t)snprintf(buffer, sizeof(buffer), "%08X", hashCode);
+    assert(written == sizeof(buffer) - 1);
+    (void)written;
+
+    return (*env)->NewStringUTF(env, buffer);
+}
+
+static bool TryGetExistingCertificateAlias(JNIEnv* env,
+                                           jobject /*KeyStore*/ store,
+                                           jobject /*X509Certificate*/ cert,
+                                           jobject* outAlias)
+{
+    // String alias = store.getCertificateAlias(cert);
+    jstring alias = (*env)->CallObjectMethod(env, store, g_KeyStoreGetCertificateAlias, cert);
+    bool containsCert = alias != NULL;
+    if (outAlias != NULL)
+    {
+        *outAlias = alias;
+    }
+    else
+    {
+        (*env)->DeleteLocalRef(env, alias);
+    }
+
+    return containsCert;
+}
+
+int32_t AndroidCryptoNative_X509StoreAddCertificate(jobject /*KeyStore*/ store, jobject /*X509Certificate*/ cert)
+{
+    assert(store != NULL);
+    assert(cert != NULL);
+
+    JNIEnv* env = GetJNIEnv();
+
+    if (TryGetExistingCertificateAlias(env, store, cert, NULL /*outAlias*/))
+    {
+        // Certificate is already in store - nothing to do
+        return SUCCESS;
+    }
+
+    jstring alias = CreateAliasForCertificate(env, cert);
+
+    // store.setCertificateEntry(alias, cert);
+    (*env)->CallVoidMethod(env, store, g_KeyStoreSetCertificateEntry, alias, cert);
+    (*env)->DeleteLocalRef(env, alias);
+
+    return CheckJNIExceptions(env) ? FAIL : SUCCESS;
+}
+
+int32_t AndroidCryptoNative_X509StoreAddCertificateWithPrivateKey(jobject /*KeyStore*/ store,
+                                                                  jobject /*X509Certificate*/ cert,
+                                                                  void* key,
+                                                                  PAL_KeyAlgorithm algorithm)
+{
+    assert(store != NULL);
+    assert(cert != NULL);
+    assert(key != NULL);
+
+    int32_t ret = FAIL;
+    JNIEnv* env = GetJNIEnv();
+
+    INIT_LOCALS(loc, alias, certs);
+
+    bool releasePrivateKey = true;
+    jobject privateKey;
+    switch (algorithm)
+    {
+        case PAL_EC:
+        {
+            EC_KEY* ec = (EC_KEY*)key;
+            privateKey = (*env)->CallObjectMethod(env, ec->keyPair, g_keyPairGetPrivateMethod);
+            break;
+        }
+        case PAL_DSA:
+        {
+            // key is a KeyPair jobject
+            privateKey = (*env)->CallObjectMethod(env, key, g_keyPairGetPrivateMethod);
+            break;
+        }
+        case PAL_RSA:
+        {
+            RSA* rsa = (RSA*)key;
+            privateKey = rsa->privateKey;
+            releasePrivateKey = false; // Private key is a global ref stored directly on RSA handle
+            break;
+        }
+        default:
+            return FAIL;
+    }
+
+    loc[alias] = CreateAliasForCertificate(env, cert);
```

Comment:
Adding just the cert looks for a current alias, but with a private key doesn't?

Reply:
This was relying on the store's behaviour of replacing existing entries (cert or private key) when setting a private key entry. That doesn't look to be the public contract of stores in general - I'll make it explicitly check like with just the cert (and remove/replace).